testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    cluster::Cluster,
+    health::{HealthCheck, HealthCheckContext},
+    prometheus::Prometheus,
+    util::unix_timestamp_now,
+};
+use anyhow::{format_err, Result};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::{collections::HashMap, time::Duration};
+
+static VAL_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"val-(\d+)").unwrap());
+static FULLNODE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"fn-(\d+)").unwrap());
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+    prometheus: Prometheus,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster, prometheus: Prometheus) -> FullNodeHealthCheck {
+        Self {
+            cluster,
+            prometheus,
+        }
+    }
+
+    pub fn state_sync_committed_version(
+        &self,
+        start: Duration,
+        end: Duration,
+    ) -> Result<HashMap<String, f64>> {
+        libra_retrier::retry(libra_retrier::fixed_retry_strategy(1_000, 5), || {
+            self.prometheus
+                .query_range_max(
+                    "libra_state_sync_committed_version{{}}".to_string(),
+                    &start,
+                    &end,
+                    10, /* step */
+                )
+                .map_err(|e| format_err!("No sync committed version data: {}", e))
+        })
+    }
+}
+
+pub fn convert_index(input: HashMap<String, f64>) -> HashMap<String, Vec<i64>> {
+    let mut res = HashMap::new();
+    for (peer_name, version) in input {
+        if let Some(cap) = VAL_REGEX.captures(peer_name.as_ref()) {
+            // this is index of validator
+            if let Some(cap) = cap.get(1) {
+                let index = cap.as_str();
+                let mut versions = vec![];
+                versions.push(-version as i64);
+                // we record version of validator to be a negative number
+                // to calculate diff with all full nodes are associated
+                res.insert(index.to_string(), versions);
+            }
+        } else if let Some(cap) = FULLNODE_REGEX.captures(peer_name.as_ref()) {
+            // this is index of fullnode
+            if let Some(cap) = cap.get(1) {
+                let index = cap.as_str();
+                if res.contains_key(index) {
+                    res.get_mut(index).unwrap().push(version as i64);
+                } else {
+                    let mut versions = vec![];
+                    versions.push(version as i64);
+                    res.insert(index.to_string(), versions);
```

Comment:
Consider using the [`entry` API](http://doc.rust-lang.org/nightly/std/collections/struct.HashMap.html#method.entry) to simplify this edition, as well as the one above.